### PR TITLE
[R-package] Fix best_iter and best_score

### LINKED ISSUE #2158: [R-package] lgb.cv returns negative best score with an absolute value metric
It seems the R package currently is able to return a negative value even when the metric is never negative, due to rounding issues. Happens especially on lgb.cv, but sometimes happens on lgb.train.

```r
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
params <- list(objective = "regression", metric = "l2")
valids <- list(test = dtest)
model <- lgb.train(params,
                   dtrain,
                   10,
                   valids,
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
```

best_score is: -6.44165e-17

### PR DESCRIPTION
**Pull request by Laurae2**
This should fix #2158 and #2029.

"best" rule:
* Take the first iteration where the best score is attained, even if the further iterations have identical scores (do not take the last)
* When there is no scoring done, the best score becomes NA and the best iteration is -1 (currently)

Later, we should enforce the metric used for early stopping should be only the first one, or at worst give the user the ability to choose the metric (best is the first).

note: @jameslamb spaces from RStudio to fix

Example (change to `metric = "auc"` and `max_depth = 3` to test maximization):

```r
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
params <- list(objective = "regression", metric = "l2")
valids <- list(test = dtest)
model <- lgb.train(params,
                   dtrain,
                   100,
                   min_data = 1,
                   learning_rate = 0.5)
model$best_score
model$best_iter
# > model$best_score
# [1] NA
# > model$best_iter
# [1] -1

model <- lgb.train(params,
                   dtrain,
                   100,
                   valids,
                   min_data = 1,
                   learning_rate = 0.5)
model$best_score
model$best_iter
# > model$best_score
# [1] 7.497024e-62
# > model$best_iter
# [1] 100

model <- lgb.train(params,
                   dtrain,
                   100,
                   valids,
                   min_data = 1,
                   learning_rate = 0.5,
                   early_stopping_rounds = 10)
model$best_score
model$best_iter
# > model$best_score
# [1] 7.497024e-62
# > model$best_iter
# [1] 100

model <- lgb.cv(params,
             dtrain,
             10,
             nfold = 5,
             min_data = 1,
             learning_rate = 1)
model$best_score
model$best_iter
# > model$best_score
# [1] 0.0003072197
# > model$best_iter
# [1] 1

model <- lgb.cv(params,
             dtrain,
             10,
             nfold = 5,
             min_data = 1,
             learning_rate = 1,
             early_stopping_rounds = 5)
model$best_score
model$best_iter
# > model$best_score
# [1] 5.369631e-18
# > model$best_iter
# [1] 3
```

### CODE REVIEW

**Comment by StrikerRUS** [comment]
> Later, we should enforce the metric used for early stopping should be only the first one, or at worst give the user the ability to choose the metric (best is the first).

There are some problems with metrics' order in Python: #2127. I suppose the same is true and for R too.

**Comment by Laurae2** [comment]
@StrikerRUS In R lists, it is fixed. The first element always remains the first element. You may even have two elements having the same exact names without any conflict.

However users may provide the same metric multiple times by mistake, in that case we deduplicate them.

Example:

```r
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
params <- list(objective = "regression", metric = c("l2", "l1", "l2"))
valids <- list(test = dtest)

model <- lgb.train(params,
                   dtrain,
                   100,
                   valids,
                   min_data = 1,
                   learning_rate = 0.5)
str(model$record_evals$test, max.level = 1)
# > str(model$record_evals$test, max.level = 1)
# List of 2
#  $ l2:List of 2
#  $ l1:List of 2
```

**Comment by guolinke** [comment]
@Laurae2 will it be better to have a parameter named "first_metric_only" in R as well?

**Comment by guolinke** [comment]
BTW, if both R and python have "first_metric_only" option, I think we should have the same option in CLI version.

**Comment by Laurae2** [comment]
@guolinke Yes, I think we should have `first_metric_only` option for CLI, R, and Python.

As the handling would be different for each wrapper, R and Python would have their own implementations using callback.

**Comment by Laurae2** [comment]
@StrikerRUS do you know why Travis MPI/Python jobs are failing?

**Comment by StrikerRUS** [comment]
@Laurae2 gcc 9 has been released recently. Hotfix is already in `master`: abbbbd7545e6a0549c17a182b7a59d7468cfb58e.

**Comment by guolinke** [comment]
> Later, we should enforce the metric used for early stopping should be only the first one, or at worst give the user the ability to choose the metric (best is the first).

@Laurae2 is this implemented in this PR?

**Comment by guolinke** [comment]
also refer to this:
https://github.com/microsoft/LightGBM/pull/2127#issuecomment-493277818
we should ensure that cli, python and R have the same behavior with `first_metric_only`.

**Comment by Laurae2** [comment]
@guolinke This PR uses all metrics. We can do `first_metric_only` in another PR.

Note that the best score / iteration is taken from the first metric when it was not computed by the model.

**Comment by Laurae2** [comment]
Closing/reopening for CI

**Comment by StrikerRUS** [comment]
@Laurae2 
>Closing/reopening for CI

This doesn't work for branch's CIs, it works only for PR's CIs.

**Comment by Laurae2** [comment]
@jameslamb do we merge it as is for the moment?

**Comment by StrikerRUS** [comment]
I've updated the branch. Now all check should be OK.
